src/testing/io.zig

```diff
@@ -0,0 +1,229 @@
+const std = @import("std");
+const os = std.os;
+const posix = std.posix;
+const mem = std.mem;
+const assert = std.debug.assert;
+const log = std.log.scoped(.io);
+
+const constants = @import("../constants.zig");
+const FIFO = @import("../fifo.zig").FIFO;
+const Time = @import("../time.zig").Time;
+const buffer_limit = @import("../io.zig").buffer_limit;
+const DirectIO = @import("../io.zig").DirectIO;
+
+/// Very simple mock IO implementation that only implements what is needed to test Storage
+pub const IO = struct {
+    /// Options for fault injection during fuzz testing
+    pub const Options = struct {
+        /// Seed for the storage PRNG.
+        seed: u64 = 0,
+
+        /// Chance out of 100 that any read will return an error.InputOutput.
+        read_fault_probability: u8 = 0,
+
+        /// Chance out of 100 that a read larger than a logical sector
+        /// will return an error.InputOutput.
+        larger_than_logical_sector_read_fault_probability: u8 = 0,
+    };
+
+    files: []const []u8,
+    options: Options,
+    prng: std.rand.DefaultPrng,
+
+    time: Time = .{},
+    completed: FIFO(Completion) = .{ .name = "io_completed" },
+
+    pub fn init(files: []const []u8, options: Options) IO {
+        return .{
+            .options = options,
+            .prng = std.rand.DefaultPrng.init(options.seed),
+            .files = files,
+        };
+    }
+
+    /// Pass all queued submissions to the kernel and peek for completions.
+    pub fn tick(io: *IO) !void {
+        while (io.completed.pop()) |completion| {
+            completion.callback(io, completion);
+        }
+    }
+
+    /// This struct holds the data needed for a single IO operation
+    pub const Completion = struct {
+        next: ?*Completion,
+        context: ?*anyopaque,
+        callback: *const fn (*IO, *Completion) void,
+        operation: Operation,
+    };
+
+    const Operation = union(enum) {
+        read: struct {
+            fd: posix.fd_t,
+            buf: [*]u8,
+            len: u32,
+            offset: u64,
+        },
+        write: struct {
+            fd: posix.fd_t,
+            buf: [*]const u8,
+            len: u32,
+            offset: u64,
+        },
+    };
+
+    /// Returned memory can be invalidated upon call to open_in_memory_file
+    fn get_file_data(io: *IO, fd: posix.fd_t) []u8 {
+        return io.files[
+            switch (posix.fd_t) {
```

Comment:
Or perhaps, we'd want our mocked `IO` to own the memory for its storage. So it could take in an allocator and sizes rather than buffers. Up to you, though!

Reply:
We can use `fd: u32` but it would require modifying all `posix.fd_t`s in the codebase to some sort of `IO.FileDescriptor` or the like. I think long-term it would make sense to do that (and maybe make all the IO in TigerBeetle use `IO` rather than the `posix` functions directly?), but not sure if it should be done in this PR.

The original implementation had an allocator and sizes but I found that this allocator-less implementation is a good simplification for our usecase that also enables us to use static allocation.

Reply:
Ah yes, I forgot that `Storage.init` takes in a `os.fd_t` and that's platform dependent :/.

Although, as you mention, I think we could get away with re-exporting `fd_t` within IO, so the testing one would have `pub const fd_t = u32;` and the others could `pub const fd_t = os.fd_t;`?

Reply:
Oh great idea, I'll do that right now then!